odoo/addons/base/tests/common.py

```diff
@@ -29,10 +29,69 @@ def setUpClass(cls):
         # Hack to use with_context and avoid manual context dict modification
         cls.env = cls.env['base'].with_context(**DISABLED_MAIL_CONTEXT).env
 
+        independent_company = cls.setup_independent_company()
+        if independent_company:
+            # avoid using the context to assign companies
+            cls.env.user.company_id = independent_company
+            cls.env.user.company_ids = [Command.set(independent_company.ids)]
+            independent_user = cls.setup_independent_user()
+            if independent_user:
+                cls.env = cls.env(user=independent_user)
+        else:
+            cls.setup_main_company()
+
+        # Make sure all class variables have the same env.
+        # Do not specify any class variables before the env changes.
+        cls.company = cls.env.company
+        cls.currency = cls.env.company.currency_id
+
         cls.partner = cls.env['res.partner'].create({
             'name': 'Test Partner',
         })
-        cls.currency = cls.env.company.currency_id
+
+        cls.group_portal = cls.env.ref('base.group_portal')
+        cls.group_user = cls.env.ref('base.group_user')
+        cls.group_system = cls.env.ref('base.group_system')
+
+    @classmethod
+    def setup_other_currency(cls, code, rates=None, **kwargs):
+        currency = cls.env['res.currency'].with_context(active_test=False).search([('name', '=', code)], limit=1)
+        currency.rate_ids.unlink()
+        if not rates:
+            rates = [('2016-01-01', 3.0), ('2017-01-01', 2.0)]
```

Comment:
I'd prefer enforcing providing rates than that magic fallback :thinking:.
Eventually make an override in the accounting common to pass those values as default.